src/Workspaces/Core/Portable/Editing/SyntaxEditor.cs

```diff
@@ -164,16 +248,29 @@ public void InsertAfter(SyntaxNode node, IEnumerable<SyntaxNode> newNodes)
         /// <param name="node">The node already existing in the tree that the new nodes will be placed after. This must be a node this is contained within a syntax list.</param>
         /// <param name="newNode">The node to place after the existing node. This node must be of a compatible type to be placed in the same list containing the existing node.</param>
         public void InsertAfter(SyntaxNode node, SyntaxNode newNode)
+            => this.InsertAfter(node, newNode, trackNewNode: false);
+
+        /// <summary>
+        /// Insert the new node after the specified node already existing in the tree.
+        /// </summary>
+        /// <param name="node">The node already existing in the tree that the new nodes will be placed after. This must be a node this is contained within a syntax list.</param>
+        /// <param name="newNode">The node to place after the existing node. This node must be of a compatible type to be placed in the same list containing the existing node.</param>
+        /// <param name="trackNewNode">Flag to indicate if the new node should be tracked for further edits.</param>
+        internal void InsertAfter(SyntaxNode node, SyntaxNode newNode, bool trackNewNode)
+            => InsertAfter(node, new[] { newNode }, trackNewNodes: trackNewNode);
+
+        private void ApplyTrackingToNodesToInsert(params SyntaxNode[] nodes)
         {
-            CheckNodeInTree(node);
-            this.InsertAfter(node, new[] { newNode });
+
         }
 
-        private void CheckNodeInTree(SyntaxNode node)
+        private void CheckNodeInOriginalTreeOrTracked(SyntaxNode node)
         {
-            if (!OriginalRoot.Contains(node))
+            if (!OriginalRoot.Contains(node) &&
+                !_allowEditsOnLazilyCreatedTrackedNewNodes &&
+                (_lazyTrackedNewNodesOpt == null || !_lazyTrackedNewNodesOpt.Contains(node)))
```

Comment:
honestly, i can't convince mysel this check is correct.

Reply:
`!_allowEditsOnLazilyCreatedTrackedNewNodes` is actually a weak check, which says the input could be a node that might be handed to us lazily from one of the prior edits which support lazy replacement nodes - I cannot think of a way to make this a stronger check unless we force complete prior edits (which adds a perf cost). Note that the syntax editor will throw an exception later when attempting to compute changed root if someone did provide an incorrect node.

`_lazyTrackedNewNodesOpt` is a strong check that input is amongst the new node added to the tree.

Reply:
i guess i'm saying: just break this into simple statements that bail out if things are ok.  but if they all don't bail out, then throw :)  then the relation is clearer and i'm not having to demorgans my way out of this :D

Reply:
Oh yes, sure.

Reply:
thanks!